### PR TITLE
[WIP] streamingccl: add polling processor to check for cutover signal

### PR DESCRIPTION
**Pull request by adityamaru**
This change adds a new polling processor to the stream ingestion flow.
This processor is currently responsible for polling the systm.jobs table
and checking if a cutover has been signalled. If it finds a signal it
uses a mirror router to propagate that signal to all the downstream
stream ingestion processors. The teardown then ensues and eventually
pushes a "cutover" meta record back to the stream ingestion resumer.

This change moves the revert range request from OnFailOrCancel, to be
invoked when the resumer gets this "cutover" meta record.

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
This is still a WIP because of the lack of tests, but I wanted to get a review to see if anything is amiss. I've also added a few inline questions I have.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go [hunk 1]
```diff
@@ -181,6 +200,23 @@ func (sip *streamIngestionProcessor) Start(ctx context.Context) context.Context
...
 	sip.eventCh = sip.merge(sip.Ctx, eventChs)
 
+	// Start listening for signals from the poller processor.
+	go func() {
+		row, meta := sip.input.Next()
+		if meta != nil {
+			if meta.Err != nil {
+				sip.MoveToDraining(nil /* err */)
+			}
+			sip.ingestionMeta = meta
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:210]: i'm not a 100% sure if I'm handling this meta correctly. Should we just have another channel (like the signalCh) where we send the meta?

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_polling_processor.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+			// Job has been signaled to complete.
+			if cutoverTime := prog.StreamIngest.CutoverTime; !cutoverTime.IsEmpty() {
+				streamIngestBytes, err := protoutil.Marshal(prog.StreamIngest)
+				if err != nil {
+					return nil, err
+				}
+				row := rowenc.EncDatumRow{
+					rowenc.DatumToEncDatum(types.Bytes, tree.NewDBytes(tree.DBytes(streamIngestBytes))),
+				}
+				return row, nil
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamingest/stream_ingestion_polling_processor.go:132]: After we signal, what is the correct behavior? Is it okay that the processor might push this same row tick seconds later? I also played around with the idea of immediately tearing down the processor once we push a row but a downstream stream ingestion processor might handle the meta before the signalled row and that doesn't seem like what we want.

Currently, the stream ingestion processor will call `ConsumerClosed()` and then teardown this processor.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go [hunk 3]
```diff
@@ -135,13 +151,15 @@ func newStreamIngestionDataProcessor(
...
 		curBatch:            make([]storage.MVCCKeyValue, 0),
 		client:              streamClient,
+		input:               input,
 		bufferedCheckpoints: make(map[streamingccl.PartitionAddress]hlc.Timestamp),
 		timer:               timeutil.NewTimer(),
+		signalCh:            make(chan rowenc.EncDatum),
 	}
 
-	if err := sip.Init(sip, post, streamIngestionResultTypes, flowCtx, processorID, output, nil, /* memMonitor */
+	if err := sip.Init(sip, post, input.OutputTypes(), flowCtx, processorID, output, nil, /* memMonitor */
```
Comment (pbardea) [inline pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:160]: Although in this case `input.OutputTypes() == streamIngestionResultsTypes` I think that this should remain as they were?

**Comment by adityamaru** [comment]
Closing in favour of #61001.

**Pull request closed without merging**